pkg/ingester/ingester.go

```diff
@@ -3764,6 +3758,38 @@ func (i *Ingester) ShutdownHandler(w http.ResponseWriter, _ *http.Request) {
 	w.WriteHeader(http.StatusNoContent)
 }
 
+// startReadRequest tries to start a read request.
+// If it was successful, startReadRequest returns a function that should be
+// called to finish the started read request once the request is completed.
+// If it wasn't successful, the causing error is returned. In this case no
+// function is returned.
+func (i *Ingester) startReadRequest() (func(error), error) {
+	start := time.Now()
+	acquiredCircuitBreakerPermit, err := i.circuitBreaker.tryAcquirePermit()
+	if err != nil {
+		return nil, err
+	}
+	err = i.checkAvailableForRead()
+	if err != nil {
+		return nil, err
+	}
+	err = i.checkReadOverloaded()
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Should we release the acquired permit here?

Reply:
That's a good catch. It was correctly handled in the initial implementation, but then it got lost while I was fixing review findings. I'll add this usecase